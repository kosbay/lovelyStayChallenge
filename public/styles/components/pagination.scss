.paginationContainer {
  display: flex;
  justify-content: flex-end;
  align-items: center;
  margin-top: 8px;

  p {
    font-size: 14px;
    color: $black;
  }

  .disabled {
    color: $lighterBlack;
  }

  .pagination {
    display: flex;
    align-items: center;
  
    li {
      display: unset;
      margin: 6px;
      font-size: 14px;
      color: $lighterBlack;
      cursor: pointer;
    }
  
    .active {
      color: $black;
    }
  }
}

.paginationArrow {
  background-color: $background;
  border: 1px solid $black;
  border-radius: 12px;
  display: flex;
  padding: 12px;
  height: 40px;
  width: 40px;

  &.disabled {
    border: 1px solid $divider;
  }

  img {
    height: 16px;
    width: 100%;
  }

  .right {
    transform: rotate(-90deg);
  }

  .left {
    transform: rotate(90deg);
  }
}