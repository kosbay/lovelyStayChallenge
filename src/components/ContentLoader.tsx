import Loader from 'react-content-loader'

interface ContentLoaderProps {}

const ContentLoader: React.FC<ContentLoaderProps> = () => {
  return (
    <Loader
      viewBox="0 0 100 100"
      style={{ width: '100%', height: '100%' }}
    >
      <rect x="0" y="5" rx="1" ry="1" width="100" height="5" />
      <rect x="0" y="13" rx="1" ry="1" width="100" height="5" />
      <rect x="0" y="21" rx="1" ry="1" width="100" height="5" />
      <rect x="0" y="29" rx="1" ry="1" width="100" height="5" />
      <rect x="0" y="37" rx="1" ry="1" width="100" height="5" />
      <rect x="0" y="45" rx="1" ry="1" width="100" height="5" />
      <rect x="0" y="53" rx="1" ry="1" width="100" height="5" />
      <rect x="0" y="61" rx="1" ry="1" width="100" height="5" />
      <rect x="0" y="69" rx="1" ry="1" width="100" height="5" />
      <rect x="0" y="77" rx="1" ry="1" width="100" height="5" />
      <rect x="50" y="86" rx="1" ry="1" width="100" height="5" />
    </Loader>
  )
}

export default ContentLoader
