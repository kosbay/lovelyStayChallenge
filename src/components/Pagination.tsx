import { useState, useEffect } from 'react'
import ReactPaginate from 'react-paginate'

export interface PaginationProps {
  pageCount: number
  currentPage: number
  count: number
  pageDataCount: number
  offset: number
  onChange: (selected: number) => void
}

const Pagination: React.FC<PaginationProps> = ({
  pageCount,
  currentPage,
  pageDataCount,
  count,
  offset,
  onChange,
}) => {
  const [selected, setSelected] = useState(0)
  const [currentPageValue, setCurrentPageValue] = useState(currentPage - 1) // Lib's indexing starts from 0

  useEffect(() => {
    setCurrentPageValue(currentPage - 1)
  }, [currentPage])

  const handlePageClick = (data: { selected: number }) => {
    if(currentPageValue !== data.selected) {
      setSelected(data.selected)
      onChange(data.selected + 1)
    }
  }

  return (
    <div className="paginationContainer">
      <p>
        {(currentPage * offset) + 1 - offset} - {(currentPage * offset) + pageDataCount - offset} from {count}
      </p>
      <ReactPaginate
        previousLabel={<p className={selected === 0 ? "disabled" : ""}>Previous</p>}
        nextLabel={<p className={selected === pageCount - 1 ? "disabled" : ""}>Next</p>}
        breakLabel="..."
        breakClassName="break-me"
        pageCount={pageCount}
        initialPage={currentPageValue}
        marginPagesDisplayed={2}
        pageRangeDisplayed={2}
        onPageChange={handlePageClick}
        containerClassName="pagination"
        activeClassName="active"
        forcePage={currentPageValue}
      />
    </div>
  )
}

export default Pagination
