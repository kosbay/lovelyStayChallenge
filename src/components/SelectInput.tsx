import { SelectHTMLAttributes } from 'react'

export type SelectInputProps = SelectHTMLAttributes<HTMLSelectElement> & {
  options: {
    label: string
    value: string
  }[]
  customClassName?: string
}

const SelectInput: React.FC<SelectInputProps> = ({
  options,
  customClassName = "",
  ...rest
}) => {
  return (
    <select className={`selectInput ${customClassName}`} {...rest}>
      {options.map(({ label, value }) => (
        <option value={value} key={value}>{label}</option>
      ))}
    </select>
  )
}

export default SelectInput;
