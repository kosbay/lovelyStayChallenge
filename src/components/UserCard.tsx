import cl from 'classnames'

interface UserCardProps {
  avatarUrl: string
  name: string
  reposTotal: number
  className?: string
}

const UserCard: React.FC<UserCardProps> = ({ avatarUrl, name, reposTotal, className }) => (
  <div className={cl("userCard", className)}>
    <div className="imgBlock">
      <img src={avatarUrl} />
    </div>
    <div className="infoBlock">
      <h2>{name}</h2>
      <p>{reposTotal} repositories</p>
    </div>
  </div>
)

export default UserCard
