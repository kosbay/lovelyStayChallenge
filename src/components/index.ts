
export { default as Table } from './Table'
export { default as Pagination } from './Pagination'
export { default as ContentLoader } from './ContentLoader'
export { default as ResultView } from './ResultView'
export { default as SelectInput } from './SelectInput'
export { default as UserCard } from './UserCard'