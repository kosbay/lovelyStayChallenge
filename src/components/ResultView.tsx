import ContentLoader from './ContentLoader'

export type ResultViewProps = {
  isError: boolean
  isNotFound: boolean
  isInitialState: boolean
  isLoading: boolean
  errorMessage?: string
}

const ResultView: React.FC<ResultViewProps> = ({
  isError,
  isNotFound,
  isInitialState,
  isLoading,
  errorMessage
}) => {
  console.log('isError', isError)
  return (
    <>
      { 
        (() => {
          if(isError) {
            return (
              <div className="resultView">
                <img src="/images/error-icon.svg" />
                <p>{errorMessage}</p>
              </div>
              
            );
          } else if(isNotFound) {
            return (
              <div className="resultView">
                <img src="/images/not-found-icon.svg" />
                <p>Result not found</p>
              </div>
            )
          } else if(isInitialState) {
            return (
              <div className="resultView">
                <img src="/images/result-view-icon.svg" />
                <p>Result will be here</p>
              </div>
            )
          } else if(isLoading) {
            return <ContentLoader />
          } else {
            return null
          }
        })()
      }
    </>
  )
}

export default ResultView
