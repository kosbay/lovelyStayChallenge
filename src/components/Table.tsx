import cl from 'classnames'

import { Pagination } from '.'
import { PaginationProps } from './Pagination'

export interface TableProps<T> extends PaginationProps {
  rows: T[]
  tableConfig: {
    rowHead: string
    rowKey: string
    isImage?: boolean
  }[]
  onRowClick?: (navigationValue: string) => void
  navigationKey?: string
}

const Table = <T extends {}>({
  tableConfig,
  rows,
  onRowClick,
  navigationKey,
  ...paginationProps
}: TableProps<T>) => (
  <div>
    <table className="plainTable">
      <thead>
        <tr>
          {tableConfig?.map(({ rowHead }) => (
            <th key={rowHead}>{rowHead}</th>
          ))}
        </tr>
      </thead>
      <tbody>
        {
          rows.map((row, key) => (
            <tr
              key={key}
              onClick={() => onRowClick && navigationKey && onRowClick(row[navigationKey])}
              className={cl(onRowClick && "clickable")}
            >
              {tableConfig.map(({ rowKey, isImage }, key) => {
                if(isImage) {
                  return <td key={key}><img src={row[rowKey]} /></td>
                }
                return (
                <td key={key}>{row[rowKey]}</td>
              )})}
            </tr>
          ))
        }
      </tbody>
    </table>
    <Pagination {...paginationProps} />
  </div>
)

export default Table
