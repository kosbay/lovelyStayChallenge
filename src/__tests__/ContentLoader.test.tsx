import renderer from 'react-test-renderer'

import { ContentLoader } from "../components"

describe('ContentLoader', () => {
  it('should render without throwing an error', () => {
    const testRenderer = renderer.create(<ContentLoader />);
    expect(testRenderer.toJSON()).toMatchSnapshot();
  });
});