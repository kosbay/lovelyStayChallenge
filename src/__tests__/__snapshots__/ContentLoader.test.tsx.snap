// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContentLoader should render without throwing an error 1`] = `
<svg
  aria-labelledby="4xzt6o-aria"
  id={null}
  role="img"
  style={
    Object {
      "height": "100%",
      "width": "100%",
    }
  }
  viewBox="0 0 100 100"
>
  <title
    id="4xzt6o-aria"
  >
    Loading...
  </title>
  <rect
    clipPath="url(#4xzt6o-diff)"
    height="100%"
    role="presentation"
    style={
      Object {
        "fill": "url(#4xzt6o-animated-diff)",
      }
    }
    width="100%"
    x="0"
    y="0"
  />
  <defs>
    <clipPath
      id="4xzt6o-diff"
    >
      <rect
        height="5"
        rx="1"
        ry="1"
        width="100"
        x="0"
        y="5"
      />
      <rect
        height="5"
        rx="1"
        ry="1"
        width="100"
        x="0"
        y="13"
      />
      <rect
        height="5"
        rx="1"
        ry="1"
        width="100"
        x="0"
        y="21"
      />
      <rect
        height="5"
        rx="1"
        ry="1"
        width="100"
        x="0"
        y="29"
      />
      <rect
        height="5"
        rx="1"
        ry="1"
        width="100"
        x="0"
        y="37"
      />
      <rect
        height="5"
        rx="1"
        ry="1"
        width="100"
        x="0"
        y="45"
      />
      <rect
        height="5"
        rx="1"
        ry="1"
        width="100"
        x="0"
        y="53"
      />
      <rect
        height="5"
        rx="1"
        ry="1"
        width="100"
        x="0"
        y="61"
      />
      <rect
        height="5"
        rx="1"
        ry="1"
        width="100"
        x="0"
        y="69"
      />
      <rect
        height="5"
        rx="1"
        ry="1"
        width="100"
        x="0"
        y="77"
      />
      <rect
        height="5"
        rx="1"
        ry="1"
        width="100"
        x="50"
        y="86"
      />
    </clipPath>
    <linearGradient
      id="4xzt6o-animated-diff"
    >
      <stop
        offset="0%"
        stopColor="#f5f6f7"
        stopOpacity={1}
      >
        <animate
          attributeName="offset"
          dur="1.2s"
          keyTimes="0; 0.25; 1"
          repeatCount="indefinite"
          values="-2; -2; 1"
        />
      </stop>
      <stop
        offset="50%"
        stopColor="#eee"
        stopOpacity={1}
      >
        <animate
          attributeName="offset"
          dur="1.2s"
          keyTimes="0; 0.25; 1"
          repeatCount="indefinite"
          values="-1; -1; 2"
        />
      </stop>
      <stop
        offset="100%"
        stopColor="#f5f6f7"
        stopOpacity={1}
      >
        <animate
          attributeName="offset"
          dur="1.2s"
          keyTimes="0; 0.25; 1"
          repeatCount="indefinite"
          values="0; 0; 3"
        />
      </stop>
    </linearGradient>
  </defs>
</svg>
`;
